Added the base for direct consumer verifier application

mmaslankaprv opened the pull request:
Added application implementation for direct consumer verifier. The direct consumer verifier is a standalone application that expose ConnectRPC API to create and manage `kafka::client::direct_consumer` instances. The application is intended to be used in ducktape to test our client implementation independently from Redpanda.

Thanks to the testing in isolation we will be able to verify the behavior of the client in scale and while injecting failures. Additionally having separate application will allow us to verify client compatibility with different implementations of kafka protocol.

NOTE:
For now the test is not being run in CI. I am preparing necessary vtools changes to make it happen
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
kc/verifier: added verifier application for direct consumer

Added simple test application that will allow basic scale testing of a
direct consumer. The application exposes REST API that the user can
interact with to drive the direct consumers to fetch from the topic.

So far the application doesn't have any way of consuming the fetched
data but it returns per partition statistics that can be used to
validate if the consumer is making progress.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--52e45be3-5bdc-422e-9214-dbd94466db50-->
<!--TEST RESULTS-->
<!--f7c7d6b1-a7bd-49d0-b392-44bfab7db0eb-->
<!--TEST RESULTS-->
<!--abcf7863-7dfc-4865-80d2-5276c2e3c2c8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69261">build#69261</a></summary>


| test_class                | test_method          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:---------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write           |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69261#01981ce3-035a-426b-ade6-ae01863eef22 | FLAKY         | 20/21    | upstream reliability is '97.70408163265306'. current run reliability is '95.23809523809523'. drift is 2.46599 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69261#01981cdd-eb78-4d7a-a368-c6d2eecb3194 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69635">build#69635</a></summary>


| test_class                 | test_method                     | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:---------------------------|:--------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests           | test_json_schema_unicode        | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/69635#01983e26-8a26-4050-a2b0-0ae44aecdfbd | FLAKY         | 20/21    | upstream reliability is '98.4567901234568'. current run reliability is '95.23809523809523'. drift is 3.21869 and the allowed drift is set to 50. The test should PASS |
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/69635#01983e1f-dbde-42fa-8d9a-c805a57e87a0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/69635#01983e26-8a21-4c85-95b7-f2a97d194bd4 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69854">build#69854</a></summary>


| test_class                  | test_method                               | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| IdempotencySnapshotDelivery | test_recovery_after_snapshot_is_delivered | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69854#019856c0-e207-45cb-851d-399aff7c4ad8 | FLAKY         | 20/21    | upstream reliability is '98.10246679316889'. current run reliability is '95.23809523809523'. drift is 2.86437 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv committed:
tests: added direct consumer verifier test

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
bazel/packaging: added `native_package` rule

Added a rule that will allow us to define a generic package containing
native seastar binaries and other binaries together in a single package.
The rule can create either a tarball or a directory.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
bazel: define directory package for direct consumer verifier

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69635-->
<!--0f3ff51e-4e7d-4d45-83df-a362ab434440-->
<!--68a6715e-8ba3-4fd2-ae35-67af65197619-->
<!--title-->
## Retry command for [Build#69635](https://buildkite.com/redpanda/redpanda/builds/69635)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

bharathv commented on src/v/kafka/client/direct_consumer/verifier/proto/verifier.proto:
```diff
@@ -0,0 +1,156 @@
+
+message AssignPartitionsRequest {
+    string client_id = 1;
+    repeated TopicAssignment topic_assignments = 3;
```
whats with these non contiguous field tags? (just curious).

rockwotj replied:
should be 2 :)

mmaslankaprv replied:
sorry, my mistake, fixing

bharathv reviewed (dismissed):
🔥  .. I'm not too familiar with the packaging parts but rest lgtm.

oleiman commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -0,0 +1,115 @@
+from rptest.tests.redpanda_test import RedpandaTest
+
+
+#TODO: This test must be enabled once the direct consumer verifier support
+# is added to vtools.
+class DirectConsumerVerifierTest(RedpandaTest):
```
is it disabled now? I don't see any `@ignore`, so how do we turn it on when ready?

mmaslankaprv replied:
It is in a separate directory i.e. it is not included in any of the test suites

oleiman reviewed (dismissed):
awesome. i also have no clue about the packaging code.

rockwotj commented on src/v/kafka/client/direct_consumer/verifier/proto/verifier.proto:
```diff
@@ -0,0 +1,156 @@
+
+message StatusRequest {}
+
+message ApiResponse {
```
nit: just use an empty proto and you can throw one of the special RPC errors to get structured errors back: https://github.com/redpanda-data/redpanda/blob/cd5407b199f2c60aa62eb829b988d4f93ff7acd6/src/v/serde/protobuf/rpc.h#L81

rockwotj commented on src/v/kafka/client/direct_consumer/verifier/proto/verifier.proto:
```diff
@@ -0,0 +1,156 @@
+service VerifierService {
+    rpc Status(StatusRequest) returns (ApiResponse) {
+        option (redpanda.pbgen.rpc) = {
+            version: V25_2_0,
+            authz: PUBLIC,
+            http_route: "/status",
```
FYI I have removed `version` from this proto

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+        "rpath_override": attr.string(mandatory = False),
+        "owner": attr.int(),
+        "install_path": attr.label(
+            default = Label("//bazel/packaging:install_path"),
```
no default for consistency

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+        "{}/lib".format(install_path_value),
+    )
+
+    binaries = []
+    for b in ctx.attr.binaries:
+        for file in b.files.to_list():
+            binaries.append(file)
```
nit: the same as `binaries = [bin.file for bin in cc_binaries]`?

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+            "source": b.path,
+        })
+
+    # in directory we put cc_binaries into libexec so they can be executed with a script that is created later
```
nit: I don't see a libexec here?

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+    for b in package_content.cc_binaries:
+        package_files.append({
+            "path": "bin",
+            "name": b.basename.removesuffix("_ld").removesuffix("_patched"),
```
How is this different than the above?

mmaslankaprv replied:
the `cc_binaries` goes through the procedure of patching them with `patchelf` whereas the other binaries do not

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+        "package_files": package_files,
+    }
+
+def _create_tarball_cfg(ctx, package_content):
```
This looks like the same as the directory one, save the bool flag. Can we unify?

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -358,3 +358,154 @@ redpanda_package = rule(
+native_package = rule(
+    implementation = _native_pkg_impl,
+    attrs = {
+        "cc_binaries": attr.label_list(),
+        "binaries": attr.label_list(),
```
Why do we need both? Can we just unify? I guess we don't run patchelf on `binaries` right? For now let's assume YAGNI and we only will need this for binaries with shared libraries.

rockwotj commented on tests/rptest/services/direct_consumer_verifier.py:
```diff
@@ -0,0 +1,308 @@
+    Convert a dataclass instance to a dictionary, excluding None values.
+    Handles nested dataclasses, lists, and enums properly.
+    """
+    if hasattr(obj, '__dataclass_fields__'):
+        # Handle dataclass objects
+        result = {}
+        for field in fields(obj):
+            value = getattr(obj, field.name)
+            if value is not None:
+                result[field.name] = as_dict(value)
+        return result
```
can't do you:

```
import dataclass
if dataclass.is_dataclass(obj):
  return dataclass.asdict(obj)
```

mmaslankaprv replied:
This way enums are incorrectly encoded

rockwotj replied:
can you just iterate and recurse over the output of asdict then?

rockwotj replied:
No but we should use `is_dataclass` instead of relying on python internal `__dataclass_fields__`

mmaslankaprv replied:
👍🏼

rockwotj commented on tests/rptest/services/direct_consumer_verifier.py:
```diff
@@ -0,0 +1,308 @@
+
+
+@dataclass
+class DirectConsumerConfiguration:
```
FYI I am working on generating protos and RPC clients for ducktape, this is fine for here however.

joe-redpanda commented on tests/rptest/services/direct_consumer_verifier.py:
```diff
@@ -0,0 +1,289 @@
+import json
+
+
+def as_dict(obj):
```
Nit: this seems useful, share this in a utils file?

joe-redpanda commented on src/v/kafka/client/direct_consumer/verifier/application.h:
```diff
@@ -0,0 +1,174 @@
+    int16_t port;
+};
+
+using request_handler_fn
```
used?

joe-redpanda commented on src/v/kafka/client/direct_consumer/verifier/application.cc:
```diff
@@ -0,0 +1,375 @@
+    auto it = _consumers.find(kafka::client_id(req.get_client_id()));
+    if (it == _consumers.end()) {
+        throw serde::pb::rpc::base_exception(
+          404,
```
we can probably yank standard http codes and messages from boost or seaster
